### DISCUSSION TITLE
Changing Address after defining

### DISCUSSION THREAD

**Discussion by andiatbaden** [orgua/OneWireHub#119] [Q&A] [answered]
I set up sensor slaves with attiny84's over onewire with the OneWire Hub Lib.

I got all to run, but i am hanging on one problem. Before the setup function i have to define the address of the sensors. The idea is to set a part of the address with Dipswitches. The function of Dipswitches is running, no problem. So it could be possible to set the address without changing code in the setup function. 

Is there any chance to redefine the address in the Setup Routine? I have not found a function in the Library to do this.

Kind regards 

andiatbaden

**Comment by orgua** [comment, marked as answer]
something like that should work, basic c tricks. hacked together and untested though! 

- first just globally declare the variable 
- later define the the variable (and also redefine it if needed)

```
#include "OneWireItem.h"
#include "OneWireHub.h"
#include "DS2405.h"

auto hub    = OneWireHub(pin_onewire);
OneWireItem ds2405;


void setup()
{
    const byte_value = read ...;
    ds2405 = DS2405( DS2405::family_code, 0x00, 0x00, 0x05, 0x24, 0xDA, byte_value );
    ...
}

...
```

**Reply by andiatbaden** [reply]
Hello,

unfortunately your solution doesn't work.

I put in your code, but the compiler threw out: 


MCP9808_OneWire_Dip_20230313:38:13: error: no matching function for call to 'OneWireItem::OneWireItem()'
In file included from D:\Dokumente\Fischgärtner\Gewächshaus\Elektronik\ArduinoSketche\MCP9808_OneWire_Dip_20230313\MCP9808_OneWire_Dip_20230313.ino:25:0:
D:\Dokumente\arduino_scripte\libraries\OneWireHub\src/OneWireItem.h:25:5: note: candidate: constexpr OneWireItem::OneWireItem(OneWireItem&&)
     OneWireItem(OneWireItem&& owItem) = default;               // default  move constructor
     ^~~~~~~~~~~
D:\Dokumente\arduino_scripte\libraries\OneWireHub\src/OneWireItem.h:25:5: note:   candidate expects 1 argument, 0 provided
D:\Dokumente\arduino_scripte\libraries\OneWireHub\src/OneWireItem.h:20:5: note: candidate: OneWireItem::OneWireItem(uint8_t, uint8_t, uint8_t, uint8_t, uint8_t, uint8_t, uint8_t)
     OneWireItem(uint8_t ID1, uint8_t ID2, uint8_t ID3, uint8_t ID4, uint8_t ID5, uint8_t ID6, uint8_t ID7);
     ^~~~~~~~~~~
D:\Dokumente\arduino_scripte\libraries\OneWireHub\src/OneWireItem.h:20:5: note:   candidate expects 7 arguments, 0 provided
MCP9808_OneWire_Dip_20230313:38:13: error: cannot declare variable 'ds2405' to be of abstract type 'OneWireItem'
In file included from D:\Dokumente\Fischgärtner\Gewächshaus\Elektronik\ArduinoSketche\MCP9808_OneWire_Dip_20230313\MCP9808_OneWire_Dip_20230313.ino:25:0:
D:\Dokumente\arduino_scripte\libraries\OneWireHub\src/OneWireItem.h:16:7: note:   because the following virtual functions are pure within 'OneWireItem':
 class OneWireItem
       ^~~~~~~~~~~
D:\Dokumente\arduino_scripte\libraries\OneWireHub\src/OneWireItem.h:44:18: note: 	virtual void OneWireItem::duty(OneWireHub*)
     virtual void duty(OneWireHub * hub) = 0;
                  ^~~~
D:\Dokumente\Fischgärtner\Gewächshaus\Elektronik\ArduinoSketche\MCP9808_OneWire_Dip_20230313\MCP9808_OneWire_Dip_20230313.ino: In function 'void setup()':
MCP9808_OneWire_Dip_20230313:53:71: error: 'byte_value' was not declared in this scope
exit status 1
no matching function for call to 'OneWireItem::OneWireItem()'

The ino file is attached.

Thank you for your support.

Kind regards

andiatbaden
[MCP9808_OneWire_Dip_20230313.ino.txt](https://github.com/orgua/OneWireHub/files/10960433/MCP9808_OneWire_Dip_20230313.ino.txt)

**Reply by orgua** [reply]
Well, at least one of the three error messages hinted you at the unknown variable "byte_value" that you re-implemented as idSwitch but kept in the code. but beside that there was another problem with virtual classes I forgot about.

Anyway, the code below compiles but is untested. Var 1 should be safe and successful to use. 

```
#include "OneWireHub.h"
#include "DS2405.h"  // Dual channel addressable switch

constexpr uint8_t pin_led       { 13 };
constexpr uint8_t pin_onewire   { 8 };

auto hub    = OneWireHub(pin_onewire);
auto ds2405a = DS2405( DS2405::family_code, 0x00, 0x00, 0x05, 0x24, 0xDA, 0x00 );

void setup()
{
    Serial.begin(115200);

    // Var 1: modifying the internal ID before attaching device
    uint8_t id1Switch = 111;
    ds2405a.ID[1] = id1Switch; // 0-based ... so ID[0] is the family-code at position 0 (see above)
    hub.attach(ds2405a);

    // Var 2: probably does not work, as the object gets destroyed, except you use "new"
    pinMode(pin_led, INPUT);
    uint8_t id2Switch = digitalRead(pin_led);
    auto ds2405b = DS2405( DS2405::family_code, 0x00, id2Switch, 0x05, 0x24, 0xDA, 0x00 );
    hub.attach(ds2405b);

    // Var 3: pointer magic
}

void loop()
{
    static bool switch_state = 0;

    // following function must be called periodically
    hub.poll();

    if (switch_state != ds2405a.getPinState())
    {
        switch_state = ds2405a.getPinState();

        // visual feedback and printF
        Serial.print(" PinState: ");
        Serial.println(switch_state);

        digitalWrite(pin_led, switch_state);
    }
}
```

**Comment by andiatbaden** [comment]
Hello Orgua,

thank you very much for your time. It is working!!

Kind regards

andiatbaden

**Comment by andiatbaden** [comment]
Sorry, variant 1 works
